src/main/java/hudson/plugins/active_directory/HttpHeaderFilter.java

```diff
@@ -0,0 +1,107 @@
+package hudson.plugins.active_directory;
+
+import hudson.model.User;
+import hudson.util.Scrambler;
+import jenkins.model.Jenkins;
+import jenkins.security.ApiTokenProperty;
+import org.acegisecurity.Authentication;
+import org.acegisecurity.GrantedAuthority;
+import org.acegisecurity.context.SecurityContextHolder;
+import org.acegisecurity.providers.UsernamePasswordAuthenticationToken;
+import org.acegisecurity.providers.anonymous.AnonymousAuthenticationToken;
+import org.acegisecurity.userdetails.UserDetails;
+import org.acegisecurity.userdetails.UsernameNotFoundException;
+
+import javax.servlet.*;
+import javax.servlet.http.HttpServletRequest;
+import java.io.IOException;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+class HttpHeaderFilter implements Filter {
+    private static final Logger LOGGER = Logger.getLogger(HttpHeaderFilter.class.getName());
+
+    private ActiveDirectorySecurityRealm activeDirectorySecurityRealm;
+
+    HttpHeaderFilter(ActiveDirectorySecurityRealm activeDirectorySecurityRealm) {
+        this.activeDirectorySecurityRealm = activeDirectorySecurityRealm;
+    }
+
+    public void doFilter(ServletRequest servletRequest,
+                         ServletResponse response, FilterChain chain)
+            throws IOException, ServletException {
+        HttpServletRequest request = (HttpServletRequest) servletRequest;
+
+        if (SecurityContextHolder.getContext().getAuthentication() == null ||
+                SecurityContextHolder.getContext().getAuthentication() instanceof AnonymousAuthenticationToken) {
+            Authentication auth = Jenkins.ANONYMOUS;
+            String authenticatedUserFromApiToken = getUserFromAuthorizationHeader(request);
+
+            String userName = authenticatedUserFromApiToken == null ? getUserFromReverseProxyHeader(request) : authenticatedUserFromApiToken;
+            if (userName != null) {
+                try {
+                    UserDetails userDetails = activeDirectorySecurityRealm.getAuthenticationProvider().loadUserByUsername(userName);
+
+                    GrantedAuthority[] authorities = userDetails.getAuthorities();
+
+                    auth = new UsernamePasswordAuthenticationToken(userName, "", authorities);
+                } catch (UsernameNotFoundException e) {
+                    LOGGER.log(Level.FINE, "User from HTTP Header {0} not found in LDAP", userName);
+                }
+            }
+
+            SecurityContextHolder.getContext().setAuthentication(auth);
```

Comment:
This should use `ACL.as()` in a try-with-resources block.

Reply:
Not sure what you mean by that. Can you give an example?
Do you mean 
```
try (ACLContext _ = ACL.as(auth)) {
   chain.doFilter(request, response);
}
```

Reply:
Yes, pretty much. `ACL.as()` is the Jenkins way of saving the current security context, setting it to the parameter, and then resetting it to the old security context in the finally block.